sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveHints.scala

```diff
@@ -102,6 +104,38 @@ object ResolveHints {
     }
   }
 
+  /**
+   * COALESCE Hint accepts name "COALESCE" and "REPARTITION".
+   * Its parameter includes a partition number.
+   */
+  class ResolveCoalesceHints(conf: SQLConf) extends Rule[LogicalPlan] {
+    private val COALESCE_HINT_NAMES = Set("COALESCE", "REPARTITION")
+
+    private def applyCoalesceHint(
+      plan: LogicalPlan,
+      numPartitions: Int,
+      shuffle: Boolean): LogicalPlan = {
+      Repartition(numPartitions, shuffle, plan)
+    }
+
+    def apply(plan: LogicalPlan): LogicalPlan = plan.resolveOperators {
+      case h: UnresolvedHint if COALESCE_HINT_NAMES.contains(h.name.toUpperCase(Locale.ROOT)) =>
+        val hintName = h.name.toUpperCase(Locale.ROOT)
+        val shuffle = hintName match {
+          case "REPARTITION" => true
+          case "COALESCE" => false
+        }
+        h.parameters match {
```

Comment:
nit: maybe we can just return `numPartitions` and then we can call `applyCoalesceHint` in order to avoid code duplication...

Reply:
Thanks for the review, @mgaido91. Don't quite get this comment. Please take a look at my latest commit.

Reply:
I meant something like:
```
val numPartitions = h.parameters  match {
  ...
}
 Repartition(numPartitions, shuffle, h.child)
```

Reply:
Ah I see. Thanks.